lib/downstream/engine.rb

```diff
@@ -6,7 +6,7 @@ module Downstream
   class Engine < ::Rails::Engine
     config.downstream = Downstream.config
 
-    config.to_prepare do
+    config.after_initialize do
```

Comment:
If we only setup subscribers during the app initialization, we lose code updates on reload—the store would still references previous versions of subscribers.

I think, what we should do here is to reset the store and then trigger the hook to let the app to register fresh versions of the subscribers.

Reply:
I agree that would be a better solution. I did some digging, but its quite hard to unsubscribe to all downstream-events since we don't keep a list of subscribers. I was able to make it work, but I'm accessing private methods, so it's very fragile.

```
config.to_prepare do
  ActiveSupport::Notifications.notifier.instance_variable_get(:@string_subscribers).each_key do |key|
    next unless key.is_a?(String) && key.starts_with?("downstream-events")

    ActiveSupport::Notifications.unsubscribe(key)
  end
  ActiveSupport.run_load_hooks("downstream-events", Downstream)
end
```

@palkan Do you have any suggestions to improve this?

Reply:
Yeah, dealing with AS internals doesn't look stable. I guess, a better solution would be to add some reset functionality to Downstream (extend the Pubsub interface: https://github.com/palkan/downstream/blob/master/lib/downstream/pubsub_adapters/stateless/pubsub.rb).